Fix asymmetric failure behavior of `Future#{zip,zipWith,traverse,sequence}` by making them fail fast regardless of ordering

lihaoyi opened the pull request:
# Problem Statement
Currently, given the following setup:

```scala
val f1 = Future{Thread.sleep(10000)}
val f2 = Future{Thread.sleep(2000); throw new Exception("Boom")}
```

The following two snippets exhibit different failure behavior:

```scala
val fa = Await.result(f1.zip(f2). Duration.Inf)
```

```scala
val fb = Await.result(f2.zip(f1). Duration.Inf)
```

`fa` fails after 10000ms, while `fb` fails after 2000ms. Both fail with `java.lang.Exception: boom`.

When zipping two `Future`s together, if the left `Future` fails early, the zipped `Future` fails early. But if the right `Future` fails early, the zipped `Future` waits until the right `Future` completes before failing.

`traverse` and `sequence` are similarly implemented with `zipWith` and should exhibit the same behavior. This all arises because `zipWith` is implemented using `flatMap`, which by definition asymmetric due to waiting fo the left `Future` to complete before even considering the right `Future`.

The current behavior makes the failure behavior of `Future`s most unpredictable; in general nobody pays attention to the order of `Future`s when zipping them together, and thus whether a `zipWith`ed/`zip`ed/`traverse`d/`sequence`d `Future` fails early or not is entirely arbitrary. 

# Fix

This PR replaces the implementation of `zipWith`, turning it from `flatMap`-based to `Promise`-based, so that when a `Future` fails early, regardless of whether it's the left or right `Future`, the resultant `Future` will fail immediately.

Implementation-wise I'm using an `AtomicReference` and `compareAndSet`, which should give us the behavior we want without any locking. It may well be possible to achieve with even less overhead, e.g. using only `volatile`s or even using no concurrency controls at all, but I couldn't come up with anything better. If anyone has a better solution I'll happily include it.

This fix would apply to all of `zip`/`zipWith`/`traverse`/`sequence`, since they all are implemented on top of `zipWith`

# Considerations

While it is possible that someone could be relying on the left-biased nature of current `zip`/`zipWith`/`traverse`/`sequence` implementation, but it seems like something that's unlikely to be reliable enough to depend upon. In my experience people generally aren't aware that `zipWith`/`zip`/`traverse`/`sequence`, and they don't generally know the total ordering of how long their Futures take to run. That means status quo behavior would just result in some `Future` fails mysterious taking longer to report for no clear reason.

Notably, the biased nature of these operators is not documented in any of their scaladoc comments.

While there is a non-zero chance that somebody could be intentionally or unintentionally depending on the biased nature of these combinators, there is a much greater chance that someone unaware of the current bias would be puzzled why their highly-concurrent system seems to be taking longer than expected in certain scenarios. It seems likely that this PR would fix more bugs than it would introduce

Note that this does not fix the left-biased fail-fast behavior of `flatMap` chains, or their equivalent `for`-comprehensions, as `flatMap`'s API is inherently left-biased. But anyone who wants fail-fast behavior can convert sections of their `flatMap` chains into `.zip`s where possible, and where not possible that's generally because there is some true data dependency between the `flatMap`s

lihaoyi committed:
Fix asymmetric failure behavior of Future#{zip,zipWith,traverse,sequence} by making them fail fast regardless of ordering

Currently, given the following setup:

```scala
val f1 = Future{Thread.sleep(10000)}
val f2 = Future{Thread.sleep(2000); throw new Exception("Boom")}
```

The following two snippets exhibit different failure behavior:

```scala
val fa = Await.result(f1.zip(f2). Duration.Inf)
```

```scala
val fb = Await.result(f2.zip(f1). Duration.Inf)
```

`fa` fails after 10000ms, while `fb` fails after 2000ms. Both fail with `java.lang.Exception: boom`.

When zipping two `Future`s together, if the left `Future` fails early, the zipped `Future` fails early. But if the right `Future` fails early, the zipped `Future` waits until the right `Future` completes before failing.

`traverse` and `sequence` are similarly implemented with `zipWith` and should exhibit the same behavior. This all arises because `zipWith` is implemented using `flatMap`, which by definition asymmetric due to waiting fo the left `Future` to complete before even considering the right `Future`.

The current behavior makes the failure behavior of `Future`s most unpredictable; in general nobody pays attention to the order of `Future`s when zipping them together, and thus whether a `zipWith`ed/`zip`ed/`traverse`d/`sequence`d `Future` fails early or not is entirely arbitrary.

This PR replaces the implementation of `zipWith`, turning it from `flatMap`-based to `Promise`-based, so that when a `Future` fails early, regardless of whether it's the left or right `Future`, the resultant `Future` will fail immediately.

Implementation-wise I'm using an `AtomicReference` and `compareAndSet`, which should give us the behavior we want without any locking. It may well be possible to achieve with even less overhead, e.g. using only `volatile`s or even using no concurrency controls at all, but I couldn't come up with anything better. If anyone has a better solution I'll happily include it.

This fix would apply to all of `zip`/`zipWith`/`traverse`/`sequence`, since they all are implemented on top of `zipWith`

While it is possible that someone could be relying on the left-biased nature of current `zip`/`zipWith`/`traverse`/`sequence` implementation, but it seems like something that's unlikely to be reliable enough to depend upon. In my experience people generally aren't aware that `zipWith`/`zip`/`traverse`/`sequence`, and they don't generally know the total ordering of how long their Futures take to run. That means status quo behavior would just result in some `Future` fails mysterious taking longer to report for no clear reason.

Notably, the biased nature of these operators is not documented in any of their scaladoc comments.

While there is a non-zero chance that somebody could be intentionally or unintentionally depending on the biased nature of these combinators, there is a much greater chance that someone unaware of the current bias would be puzzled why their highly-concurrent system seems to be taking longer than expected in certain scenarios. It seems likely that this PR would fix more bugs than it would introduce

Note that this does not fix the left-biased fail-fast behavior of `flatMap` chains, or their equivalent `for`-comprehensions, as `flatMap`'s API is inherently left-biased. But anyone who wants fail-fast behavior can convert sections of their `flatMap` chains into `.zip`s where possible, and where not possible that's generally because there is some true data dependency between the `flatMap`s

lihaoyi-databricks commented:
/rebuild

viktorklang commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -413,8 +413,29 @@ trait Future[+T] extends Awaitable[T] {
+      case scala.util.Success(v) =>
+        val isFirst = buffer.compareAndSet(null, Left(v))
+        if (!isFirst) {
+          outPromise.success(f(v, buffer.get.getOrElse(throw new Exception("CANT HAPPEN"))))
```
if `f` throws here then that exception will not propagate to `outPromise`.

viktorklang commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -413,8 +413,29 @@ trait Future[+T] extends Awaitable[T] {
+      case scala.util.Success(v) =>
+        val isFirst = buffer.compareAndSet(null, Right(v))
+        if (!isFirst) {
+          outPromise.success(f(buffer.get.swap.getOrElse(throw new Exception("CANT HAPPEN")), v))
```
if `f` throws here then that exception will not propagate to `outPromise`.

viktorklang commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -413,8 +413,29 @@ trait Future[+T] extends Awaitable[T] {
    */
-  def zipWith[U, R](that: Future[U])(f: (T, U) => R)(implicit executor: ExecutionContext): Future[R] =
-    flatMap(r1 => that.map(r2 => f(r1, r2)))(if (executor.isInstanceOf[BatchingExecutor]) executor else parasitic)
+  def zipWith[U, R](that: Future[U])(f: (T, U) => R)(implicit executor: ExecutionContext): Future[R] = {
```
@lihaoyi Have you benchmarked this change?

lihaoyi-databricks replied:
@viktorklang I haven't benchmarked the per-call overhead, but I have verified that the updated logic it stops slow futures from blocking other future failures from being immediately reported, which shaves 10s of seconds off some of my workflows.

viktorklang replied:
@lihaoyi It's important to make sure that it does not have an undue impact on other logic by reducing the baseline performance for everyone. you can probably use something like this: https://github.com/scala/scala/blob/2.13.x/test/benchmarks/src/main/scala/scala/concurrent/FutureBenchmark.scala#L232

lihaoyi replied:
@viktorklang got it. Is there any convenient command I can run to run those? Or docs on how to run JMH suites? I'm not that familiar with the benchmarking infra in the scala/scala build

lihaoyi replied:
`sbt "bench/jmh:runMain scala.concurrent.ZipWithFutureBenchmark"` is giving me `Error: Main method not found in class scala.concurrent.ZipWithFutureBenchmark`

lihaoyi replied:
scratch that, `sbt "bench/jmh:run scala.concurrent.ZipWithFutureBenchmark"` seems to work

lihaoyi commented:
@viktorklang I wrapped the calls to `f()` in `scala.util.Try`s; I think that should handle exceptions and put them into `outPromise` correctly?

I also converted the `.failure` calls to `.tryFailure`, since if both futures fail we don't want to trigger the `outPromise` twice. I don't think we need to change the `.complete` calls in the same way, since those only trigger if `!isFirst`, meaning that the other `Future` should have completed with a Success, `f` has not been run, and thus nobody else could have put anything into `outPromise` yet

julienrf commented:
Relates to #8994

viktorklang commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -413,8 +413,31 @@ trait Future[+T] extends Awaitable[T] {
+      case scala.util.Failure(e) => outPromise.tryFailure(e)
+    }
+    outPromise.future
+  }
```
@lihaoyi I think it could be possible to improve on the performance by avoiding wrapping, and avoiding compare-and-set.

```scala
  def zipWith[U, R](that: Future[U])(f: (T, U) => R)(implicit executor: ExecutionContext): Future[R] = {
      val buffer = new AtomicReference[Success[Any]]()
      val zipped = Promise[R]()
      this.onComplete {
        case left: Success[T] =>
          val right = buffer.getAndSet(left).asInstanceOf[Success[U]]
          if (right ne null)
            zipped.tryComplete(left.map(l => f(l, right.get)))
        case f: Failure[T] =>
          zipped.tryComplete(f.asInstanceOf[Failure[R]])
      }
      that.onComplete {
        case right: Success[U] =>
          val left = buffer.getAndSet(right).asInstanceOf[Success[T]]
          if (left ne null)
            zipped.tryComplete(right.map(r => f(left.get, r)))
        case f: Failure[U] =>
          zipped.tryComplete(f.asInstanceOf[Failure[R]])
      }
      zipped.future
    }
```

viktorklang replied:
@lihaoyi The solution above is correct, if you want to get rid of the allocation if the function(s) to `map` it is safe to do:

```scala
  def zipWith[U, R](that: Future[U])(f: (T, U) => R)(implicit executor: ExecutionContext): Future[R] = {
      val buffer = new AtomicReference[Success[Any]]()
      val zipped = Promise[R]()
      this.onComplete {
        case left: Success[T] =>
          val right = buffer.getAndSet(left).asInstanceOf[Success[U]]
          if (right ne null)
            zipped.tryComplete(try Success(f(left.get, right.get)) catch { case e if NonFatal(e) => Failure(e) })
        case f: Failure[T] =>
          zipped.tryComplete(f.asInstanceOf[Failure[R]])
      }
      that.onComplete {
        case right: Success[U] =>
          val left = buffer.getAndSet(right).asInstanceOf[Success[T]]
          if (left ne null)
            zipped.tryComplete(try Success(f(left.get, right.get)) catch { case e if NonFatal(e) => Failure(e) })
        case f: Failure[U] =>
          zipped.tryComplete(f.asInstanceOf[Failure[R]])
      }
      zipped.future
    }
```

lihaoyi replied:
I think I ended up with more or less the same thing, except I used `complete` instead of `tryComplete` for the success-with-previous-value case case, since AFAICT that case can never happen when the `outPromise` has already been completed

viktorklang replied:
@lihaoyi 

this is de default implementation of `complete`: 

```scala
  def complete(result: Try[T]): this.type =
    if (tryComplete(result)) this else throw new IllegalStateException("Promise already completed.")
```

So it is faster to just call tryComplete directly since it will not fail.

lihaoyi commented:
@viktorklang I ran `sbt "bench/jmh:run scala.concurrent.ZipWithFutureBenchmark -wi 10 -i 10"` on my laptop (2015 Macbook Pro) running OSX 10.14.3 openjdk version "11.0.7" 2020-04-14

Numbers on this PR:

```
[info] Benchmark                    (pool)  (recursion)  (threads)   Mode  Cnt   Score   Error   Units
[info] ZipWithFutureBenchmark.post     fjp         1024          1  thrpt   10   4.834 ± 0.014  ops/ms
[info] ZipWithFutureBenchmark.post     fix         1024          1  thrpt   10   2.575 ± 0.014  ops/ms
[info] ZipWithFutureBenchmark.post     fie         1024          1  thrpt   10   8.012 ± 0.393  ops/ms
[info] ZipWithFutureBenchmark.post     gbl         1024          1  thrpt   10   4.758 ± 0.058  ops/ms
[info] ZipWithFutureBenchmark.pre      fjp         1024          1  thrpt   10   6.213 ± 0.015  ops/ms
[info] ZipWithFutureBenchmark.pre      fix         1024          1  thrpt   10   2.911 ± 0.003  ops/ms
[info] ZipWithFutureBenchmark.pre      fie         1024          1  thrpt   10  10.087 ± 0.023  ops/ms
[info] ZipWithFutureBenchmark.pre      gbl         1024          1  thrpt   10   6.507 ± 0.121  ops/ms
```

Numbers on 2.13.x

```
[info] Benchmark                    (pool)  (recursion)  (threads)   Mode  Cnt   Score   Error   Units
[info] ZipWithFutureBenchmark.post     fjp         1024          1  thrpt   10   6.927 ± 0.087  ops/ms
[info] ZipWithFutureBenchmark.post     fix         1024          1  thrpt   10   6.897 ± 0.123  ops/ms
[info] ZipWithFutureBenchmark.post     fie         1024          1  thrpt   10   8.300 ± 0.013  ops/ms
[info] ZipWithFutureBenchmark.post     gbl         1024          1  thrpt   10   4.925 ± 0.076  ops/ms
[info] ZipWithFutureBenchmark.pre      fjp         1024          1  thrpt   10   7.469 ± 0.202  ops/ms
[info] ZipWithFutureBenchmark.pre      fix         1024          1  thrpt   10   7.693 ± 0.304  ops/ms
[info] ZipWithFutureBenchmark.pre      fie         1024          1  thrpt   10  11.474 ± 0.024  ops/ms
[info] ZipWithFutureBenchmark.pre      gbl         1024          1  thrpt   10   6.222 ± 0.037  ops/ms
```

Not quite sure what to make of this, or what the various `pre` `post` `fjp` `fix` `fie` `gbl` labels mean. It seems that this PR is moderately slower? Is that an OK amount, or are there things we should optimize here? e.g. we can probably remove a bunch of boxing in `Either`s or `Try`s if we really wanted, and maybe do something with that `parasitic` EC, but I'm not the expert in performance-optimizing this stuff

viktorklang commented:
@lihaoyi 

post = Futures are completed post-operation
pre = Futures are completed pre-operation

(pool) is what thread pool is being used. fjp = ForkJoinPool, fix = a fixed-size ThreadPool, fie = Future.InternalExecutionContext (parasitic), gbl = ExecutionContext.global

TBH those performance degradations are substantial, some of them are almost 30% of the original performance numbers.

lihaoyi commented:
Here's my latest benchmarks:

```
[info] ZipWithFutureBenchmark.post     fjp         1024          1  thrpt   10   4.933 ± 0.010  ops/ms
[info] ZipWithFutureBenchmark.post     fix         1024          1  thrpt   10   2.647 ± 0.026  ops/ms
[info] ZipWithFutureBenchmark.post     fie         1024          1  thrpt   10   9.130 ± 0.033  ops/ms
[info] ZipWithFutureBenchmark.post     gbl         1024          1  thrpt   10   7.341 ± 0.026  ops/ms
[info] ZipWithFutureBenchmark.pre      fjp         1024          1  thrpt   10   6.448 ± 0.012  ops/ms
[info] ZipWithFutureBenchmark.pre      fix         1024          1  thrpt   10   3.087 ± 0.006  ops/ms
[info] ZipWithFutureBenchmark.pre      fie         1024          1  thrpt   10  10.880 ± 0.015  ops/ms
[info] ZipWithFutureBenchmark.pre      gbl         1024          1  thrpt   10   8.769 ± 0.125  ops/ms
```

viktorklang commented:
@lihaoyi From a semantics PoV, the scaladoc says:

>   *  If `this` future fails, the resulting future is failed
>    *  with the throwable stored in `this`.
>    *  Otherwise, if `that` future fails, the resulting future is failed
>    *  with the throwable stored in `that`.

Now, changing so that `that`s exception is set instead of `this`s may cause issues if people are relying on that fact.
Some might. Ultimately that change of the semantics needs to be decided by the Scala team: @lrytz?

As for performance, I think it should be possible to implement some optimizations in impl/Promise.scala (DefaultPromise) in an override.

lihaoyi commented:
That's a good point; the scaladoc does hint at the left future taking precedence for fail-fast behavior, but it's not really explicit about it.

Agreed that this is a behavioral change, as I mentioned in the PR description. It's up to the maintainers to decide what to do with it: get it in now, keep it for some future major version bump where such a behavioral change, or not do it at all.

lrytz commented:
I think this makes sense since `zip`/`traverse`/`sequence` run the futures in parallel. Side effects of individual futures happen in any order, and it seems reasonable to me that the returned failure, which is due to a side effect, is not deterministic.

A counter-argument: currently a `zip`/`traverse`/`sequence` future only completes after all of its inputs have completed. It's possible that this is used as a synchronization mechanism.

lrytz commented:
> currently a `zip`/`traverse`/`sequence` future only completes after all of its inputs have completed

Ah, no this is not true. The `zip` future currently completes as soon as its first input has failed.

lihaoyi commented:
> I think this makes sense since `zip`/`traverse`/`sequence` run the futures in parallel. Side effects of individual futures happen in any order, and it seems reasonable to me that the returned failure, which is due to a side effect, is not deterministic.

I'm not sure if this is the way I would put it. `zip`/etc. don't run the futures at all, but merely create a new future that waits on them. I guess if you do `Future{...}.zip(Future{...})` it would run them in parallel, but that's more a property of `Future{...}` rather than a property of `zip`.

> Ah, no this is not true. The zip future currently completes as soon as its first input has failed.

Yeah, the fact these operators complete as soon as the left-most input fails is what I find surprising. I expect most people would expect it to be symmetric, so either "fail as soon as first future fails (time-wise)" or "complete after all futures complete" like how the success behavior works. I don't expect most people were aware that the left-most inputs are fail-fast while the right-most inputs are fails-low; I know I didn't, it seems @lrytz didn't either, and I suspect if we poll people who've been using Scala for years they mostly won't know. @sjrd in the gitter channel said he would expect the new behavior.

@viktorklang is right that this is a behavioral change, so we should tread carefully; people could be depending on this behavior accidentally whether they were aware of it or not. Despite that, I do think that the new behavior is sufficiently more "correct" that it's worth considering, and that changing this behavior has as much potential to fix old bugs as it does to create new ones.

julienrf reviewed (approved):
I’m in favor of this behavioral change.

viktorklang commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -413,8 +413,37 @@ trait Future[+T] extends Awaitable[T] {
+    val buffer = new AtomicReference[AnyRef](sentinel)
+    val outPromise = Promise[R]()
+
+    val myExecutor = if (executor.isInstanceOf[BatchingExecutor]) executor else parasitic
```
@lihaoyi This is wrong—it would make `f` evaluated on the current thread if `this` or `that` has already been completed when the callback is added. It *must* be executed on the passed in EC. If you look at the default implementation of `zipWith` (which this PR is replacing) you can see that the choice of doing `if (executor.isInstanceOf[BatchingExecutor]) executor else parasitic` only applies to the `flatMap` and not the `map`: https://github.com/scala/scala/pull/9655/files#diff-1d7e8b347eefe51392356ff23a3a480317df5d42fc10d12336c344bd1407dc75L417

viktorklang commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -413,8 +413,37 @@ trait Future[+T] extends Awaitable[T] {
+        val prevValue = buffer.getAndSet(v.asInstanceOf[AnyRef])
+        if (sentinel ne prevValue) {
+          val oldValue = prevValue.asInstanceOf[U]
+          outPromise.complete(try Success(f(v, oldValue)) catch{case e: Exception => Failure(e)})
```
We need `case e if NonFatal(e) =>`.

viktorklang commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -413,8 +413,37 @@ trait Future[+T] extends Awaitable[T] {
+        val prevValue = buffer.getAndSet(v.asInstanceOf[AnyRef])
+        if (sentinel ne prevValue) {
+          val oldValue = prevValue.asInstanceOf[T]
+          outPromise.complete(try Success(f(oldValue, v)) catch{case e: Exception => Failure(e)})
```
We need `case e if NonFatal(e) =>`.

viktorklang commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -413,8 +413,37 @@ trait Future[+T] extends Awaitable[T] {
+
+    this.onComplete{
+      case scala.util.Success(v) =>
+        val prevValue = buffer.getAndSet(v.asInstanceOf[AnyRef])
```
```suggestion
        val prevValue = buffer.getAndSet(v.asInstanceOf[R])
```

viktorklang commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -413,8 +413,37 @@ trait Future[+T] extends Awaitable[T] {
-  def zipWith[U, R](that: Future[U])(f: (T, U) => R)(implicit executor: ExecutionContext): Future[R] =
-    flatMap(r1 => that.map(r2 => f(r1, r2)))(if (executor.isInstanceOf[BatchingExecutor]) executor else parasitic)
+  def zipWith[U, R](that: Future[U])(f: (T, U) => R)(implicit executor: ExecutionContext): Future[R] = {
+    val sentinel = new Object()
```
The sentinel is not needed if you use the logic I proposed in https://github.com/scala/scala/pull/9655/files#r645659226 and thus saves an allocation.

lihaoyi commented:
Updated in response to feedback. Also removed two asserts from `FutureTests` that no long apply since `zipWith` now always create a new `Future` in all success and failure cases

viktorklang commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -413,8 +413,34 @@ trait Future[+T] extends Awaitable[T] {
+        val prevValue = buffer.getAndSet(s)
+        if (prevValue ne null) {
+          val oldValue = prevValue.get.asInstanceOf[U]
+          outPromise.complete(try Success(f(s.get, oldValue)) catch{case NonFatal(e) => Failure(e)})
```
Using the extractor for NonFatal is not optimal for performance, so use the guard: `catch { case e if NonFatal(e) => Failure(e) }`

lihaoyi replied:
got it, I changed it to a guard

viktorklang commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -413,8 +413,34 @@ trait Future[+T] extends Awaitable[T] {
-    flatMap(r1 => that.map(r2 => f(r1, r2)))(if (executor.isInstanceOf[BatchingExecutor]) executor else parasitic)
+  def zipWith[U, R](that: Future[U])(f: (T, U) => R)(implicit executor: ExecutionContext): Future[R] = {
+    val buffer = new AtomicReference[Success[Any]]()
+    val outPromise = Promise[R]()
```
@lihaoyi The main drawback of this solution is that now we tye a specific Promise-implementation into the Future-trait, which has been avoided thus far, as such other implementations of Future may now silently switch to a different implementation mid-transformation. Worth discussing

lihaoyi replied:
I didn't even know you could have other implementations of `Future`; I've never seen or used one personally. Presumably if we had such other implementations, they could override `zipWith` to get whatever behavior they wanted?

Looking at the other "fundamental" methods of future, `def transform` and `def transformWith`, they are basically defined to be "linear" dataflows, so they almost by-definition cannot handle this "whoever completes first" logic that `.zipWith` now provides. In many ways, this new logic is similar to the logic in `Future.firstCompletedOf`, which also uses promises and would exhibit the behavior you describe (feed custom `Future`s in, get non-custom `Future`s out)

If you'd like, we could turn `def zipWith` into a third "fundamental" method, leaving it abstract in `trait Future` and moving the implementation into `DefaultPromise. That may better fit it into the current framework of how `Future` methods are laid out, with a small set of abstract core methods and all other methods just building on top of these.

viktorklang replied:
@lihaoyi That's a great point. A good "middle-ground" which would be binary-backwards compatible, would be to leave the implementation in `Future` as is, and then overriding the default implementation in DefaultPromise with the new-updated-behavior. Having the implementation in `DefaultPromise` would also allow us to perform additional optimizations since we know strictly more about `this`.

viktorklang replied:
@lihaoyi If we leave the implementation in Future as-is, we can do something like this in DefaultPromise:

```scala
    override final def zipWith[U, R](that: Future[U])(f: (T, U) => R)(implicit executor: ExecutionContext): Future[R] = {
      val state = get()
      if (state.isInstanceOf[Failure[T]]) {
        this.asInstanceOf[Future[R]]
      } else if (state.isInstanceOf[Success[T]]) {
        val l = state.asInstanceOf[Success[T]].get
        that.map(r => f(l, r))
      } else {
        val buffer = new AtomicReference[Success[Any]]()
        val zipped = new DefaultPromise[R]()
        this.onComplete {
          case left: Success[T] =>
            val right = buffer.getAndSet(left).asInstanceOf[Success[U]]
            if (right ne null)
              zipped.tryComplete(try Success(f(left.get, right.get)) catch { case e if NonFatal(e) => Failure(e) })
          case f: Failure[T] =>
            zipped.tryComplete(f.asInstanceOf[Failure[R]])
        }
        that.onComplete {
          case right: Success[U] =>
            val left = buffer.getAndSet(right).asInstanceOf[Success[T]]
            if (left ne null)
              zipped.tryComplete(try Success(f(left.get, right.get)) catch { case e if NonFatal(e) => Failure(e) })
          case f: Failure[U] =>
            zipped.tryComplete(f.asInstanceOf[Failure[R]])
        }
        zipped.future
      }
    }
```

lihaoyi commented:
@viktorklang I updated the PR with your suggested implementation, and also added a simple unit test to validate the new behavior.

```
[info] Benchmark                    (pool)  (recursion)  (threads)   Mode  Cnt   Score   Error   Units
[info] ZipWithFutureBenchmark.post     fjp         1024          1  thrpt   10   4.937 ± 0.012  ops/ms
[info] ZipWithFutureBenchmark.post     fix         1024          1  thrpt   10   2.661 ± 0.007  ops/ms
[info] ZipWithFutureBenchmark.post     fie         1024          1  thrpt   10   9.135 ± 0.463  ops/ms
[info] ZipWithFutureBenchmark.post     gbl         1024          1  thrpt   10   4.953 ± 0.098  ops/ms
[info] ZipWithFutureBenchmark.pre      fjp         1024          1  thrpt   10   6.611 ± 0.057  ops/ms
[info] ZipWithFutureBenchmark.pre      fix         1024          1  thrpt   10   3.098 ± 0.005  ops/ms
[info] ZipWithFutureBenchmark.pre      fie         1024          1  thrpt   10  22.761 ± 0.085  ops/ms
[info] ZipWithFutureBenchmark.pre      gbl         1024          1  thrpt   10   6.679 ± 0.055  ops/ms
```

Generally faster than what I had before. Some things are still slower, some things are faster, and somehow `pre      fie` is now twice as fast as on `2.13.x`

lihaoyi commented:
Looks like I'm getting MIMA binary compatibility check failures. Anyone here can help me understand it? I'm not familiar with MIMA

```

[error] scala-library: Failed binary compatibility check against org.scala-lang:scala-library:2.13.6! Found 1 potential problems
[error]  * method zipWith(scala.concurrent.Future,scala.Function2,scala.concurrent.ExecutionContext)scala.concurrent.Future in class scala.concurrent.impl.Promise#DefaultPromise is declared final in current version
[error]    filter with: ProblemFilters.exclude[FinalMethodProblem]("scala.concurrent.impl.Promise#DefaultPromise.zipWith")

```

viktorklang commented:
@lihaoyi 

>Generally faster than what I had before. Some things are still slower, some things are faster, 

I think there's still room for improvement, I'll have a look.

>and somehow pre fie is now twice as fast as on 2.13.x

That's rather unsurprising since it is reduced from a flatMap + map to just a map for pre-completed operations, and `fie` runs on calling thread so it should be very fast.

viktorklang commented:
@lihaoyi 

>Looks like I'm getting MIMA binary compatibility check failures. Anyone here can help me understand it? I'm not familiar with MIMA

That's because DefaultPromise used to use the non-final version of zipWith provided by the Future-trait, and now we override it final-ly. This should be safe since DefaultPromise is a standard-library-internal implementation, but it needs to get a MIMA filter applied so it won't error the build.

viktorklang commented on test/files/jvm/future-spec/FutureTests.scala:
```diff
@@ -137,7 +137,6 @@ class FutureTests extends MinimalScalaTest {
       val f = Future.failed[Future[String]](e)
 
       assert(f.mapTo[String] eq f, "Future.mapTo must be the same instance as Future.mapTo")
-      assert(f.zip(f) eq f, "Future.zip must be the same instance as Future.zip")
```
This should be possible to be reinstated now

viktorklang commented on test/junit/scala/concurrent/FutureTest.scala:
```diff
@@ -6,8 +6,47 @@ import org.junit.Test
+    // Make sure that it fails early, after the earlier failure occurs, and does not wait
+    // for the later failure regardless of which one is on the left and which is on the right
+    assert(end - start > 150)
+    assert(end - start < 950)
```
I think that there's a risk this will error out on slow machines.
Also, do not use `currentTimeMillis` for checking elapsed time, use `nanoTime` as `currentTimeMillis` has limited accuracy on many OS:es

viktorklang commented on src/library/scala/concurrent/impl/Promise.scala:
```diff
@@ -130,6 +130,35 @@ private[concurrent] object Promise {
+        }
+        zipped.future
+      }
+    }
```
The following should be slightly faster:

```scala
override final def zipWith[U, R](that: Future[U])(f: (T, U) => R)(implicit executor: ExecutionContext): Future[R] = {
      val state = get()
      if (state.isInstanceOf[Try[T]]) {
        if (state.asInstanceOf[Try[T]].isFailure) this.asInstanceOf[Future[R]]
        else {
          val l = state.asInstanceOf[Success[T]].get
          that.map(r => f(l, r))
        }
      } else {
        val buffer = new AtomicReference[Success[Any]]()
        val zipped = new DefaultPromise[R]()

        val thisF: Try[T] => Unit = {
          case left: Success[T] =>
            val right = buffer.getAndSet(left).asInstanceOf[Success[U]]
            if (right ne null)
              zipped.tryComplete(try Success(f(left.get, right.get)) catch { case e if NonFatal(e) => Failure(e) })
          case f => // Can only be Failure
            zipped.tryComplete(f.asInstanceOf[Failure[R]])
        }

        val thatF: Try[U] => Unit = {
          case right: Success[U] =>
            val left = buffer.getAndSet(right).asInstanceOf[Success[T]]
            if (left ne null)
              zipped.tryComplete(try Success(f(left.get, right.get)) catch { case e if NonFatal(e) => Failure(e) })
          case f => // Can only be Failure
            zipped.tryComplete(f.asInstanceOf[Failure[R]])
        }
        // Cheaper than this.onComplete since we already polled the state
        this.dispatchOrAddCallbacks(state, new Transformation[T, Unit](Xform_onComplete, thisF, executor))
        that.onComplete(thatF)
        zipped.future
      }
    }
```

lihaoyi replied:
@viktorklang I notice we have a fast path if the LHS `this` future in the `zip` has already completed; should we do fast paths for other combinations of things being already completed? e.g.

- LHS completed, RHS not
- RHS completed, LHS not
- RHS completed, LHS completed

We'd then only need to fall back to the `AtomicReference`/`DefaultPromise` implementation in the case where both futures aren't completed at time of `.zip`

lihaoyi replied:
Something like this

```scala
    override final def zipWith[U, R](that: Future[U])(f: (T, U) => R)(implicit executor: ExecutionContext): Future[R] = {
      (get(), that.value) match{
        case (_: Failure[T], _) => this.asInstanceOf[Future[R]]
        case (_, Some(_: Failure[U])) => that.asInstanceOf[Future[R]]
        case (t: Success[T], Some(r: Success[U])) => Future(f(t.value, r.value))
        case (t: Success[T], None) => that.map(f(t.value, _))
        case (null, Some(r: Success[U])) => this.map(f(_, r.value))
        case (_, _) =>
          val buffer = new AtomicReference[Success[Any]]()
          val zipped = new DefaultPromise[R]()
          this.onComplete {
            case left: Success[T] =>
              val right = buffer.getAndSet(left).asInstanceOf[Success[U]]
              if (right ne null)
                zipped.tryComplete(try Success(f(left.get, right.get)) catch { case e if NonFatal(e) => Failure(e) })
            case f: Failure[T] =>
              zipped.tryComplete(f.asInstanceOf[Failure[R]])
          }
          that.onComplete {
            case right: Success[U] =>
              val left = buffer.getAndSet(right).asInstanceOf[Success[T]]
              if (left ne null)
                zipped.tryComplete(try Success(f(left.get, right.get)) catch { case e if NonFatal(e) => Failure(e) })
            case f: Failure[U] =>
              zipped.tryComplete(f.asInstanceOf[Failure[R]])
          }
          zipped.future
      }
    }
```

```scala

    // If either of the upstream futures has already failed by the time we call these operators,
    // we just re-use it and don't allocate anything
    val f3 = Future.failed(new Exception())
    val f4 = Future {Thread.sleep(1000)}

    assert(f3.zip(f4) eq f3)
    assert(f3.zipWith(f4)((_, _)) eq f3)
    assert(f4.zip(f3) eq f3)
    assert(f4.zipWith(f3)((_, _)) eq f3)

    // Exercise the "both futures were already successful when we call the combinator" fast path
    val f5 = Future.successful(123)
    val f6 = Future.successful(456)
    assert(Await.result(f5.zip(f6), Inf) == (123, 456))
    assert(Await.result(f6.zip(f5), Inf) == (456, 123))
    assert(Await.result(f5.zipWith(f6)((_, _)), Inf) == (123, 456))
    assert(Await.result(f6.zipWith(f5)((_, _)), Inf) == (456, 123))

    // Exercise the "either future was already successful when we call the combinator" fast path
    val f7 = Future.successful(123)
    val f8 = Future{Thread.sleep(1000); 456}
    assert(Await.result(f7.zip(f8), Inf) == (123, 456))
    assert(Await.result(f8.zip(f7), Inf) == (456, 123))
    assert(Await.result(f7.zipWith(f8)((_, _)), Inf) == (123, 456))
    assert(Await.result(f8.zipWith(f7)((_, _)), Inf) == (456, 123))
```

lihaoyi replied:
I'm not sure how much we care about the "already completed" fast paths here, but I figure if we're gonna fast path one special case we should fast path all three of them (especially given how little code is required to do so)

viktorklang commented:
I purposefully didn’t do that because we cannot make assumptions w.r.t. the
performance of calling ’value’ on a generic Future, in the DefaultPromuse
implementation it needs to allocate a Some. Allocating a Tuple2 and making
additional type tests and extractors doesn’t make much sense given that
we’ll call ’map’ if we are a Success already, and if you look at the
implementation of ’map’ for DefaultPromuse you can see that it immediate
returns itself if it is already Failure which will be much faster that
allocating Tuples and running extractors. For a generic Future it must be
presumed that ’map’ is a very common operation.--
Cheers,
√

lihaoyi commented:
ok, happy to go with your code then

lihaoyi commented:
Updated the PR to use @viktorklang's suggested implementation, and made the test suite use `nanoTime`.

viktorklang commented:
@lihaoyi I think you need to add the following to make MIMA happy:

```diff
diff --git project/MimaFilters.scala project/MimaFilters.scala
index 0b35213fff..34ba4e3d3c 100644
--- project/MimaFilters.scala
+++ project/MimaFilters.scala
@@ -39,6 +39,9 @@ object MimaFilters extends AutoPlugin {
     ProblemFilters.exclude[DirectMissingMethodProblem]("scala.Predef#SeqCharSequence.isEmpty"),
     ProblemFilters.exclude[DirectMissingMethodProblem]("scala.Predef#ArrayCharSequence.isEmpty"),
     ProblemFilters.exclude[DirectMissingMethodProblem]("scala.runtime.ArrayCharSequence.isEmpty"),
+
+    // this is an internal class and adding a final override here should not be a problem
+    ProblemFilters.exclude[FinalMethodProblem]("scala.concurrent.impl.Promise#DefaultPromise.zipWith")
   )

   override val buildSettings = Seq(
```

lihaoyi commented:
Cool lemme try that

lihaoyi commented:
The timing-based test case is a bit clunky, but I suppose we are testing timing/ordering-based things, so hopefully it'll be stable enough.

The other alternative is to rewrite the test using raw `Promise`s where we can manually control the order of completion; doable, but IMO less clear at-a-glance than the current test code using `Thread.sleep`:

```diff
@@ -13,9 +13,13 @@ class FutureTest {
   def testZipWithFailFastBothWays(): Unit = {
     import ExecutionContext.Implicits.global

-    val start = System.nanoTime()
-    val f1 = Future{Thread.sleep(1000); throw new Exception("Boom Late"); 123}
-    val f2 = Future{Thread.sleep(200); throw new Exception("Boom Early"); 456}
+    val p1 = Promise[Int]()
+    val p2 = Promise[Int]()
+    // Make sure that it fails early, after the earlier failure occurs, and does not wait
+    // for the later failure regardless of which one is on the left and which is on the right
+    p1.failure(new Exception("Boom Early"))
+    val f1 = p1.future
+    val f2 = p2.future

     val scala.util.Failure(fa) = Try(Await.result(f1.zip(f2), Inf))
     val scala.util.Failure(fb) = Try(Await.result(f2.zip(f1), Inf))
@@ -29,7 +33,6 @@ class FutureTest {
     val scala.util.Failure(fg) = Try(Await.result(Future.traverse(Seq(0, 1))(Seq(f1, f2)(_)), Inf))
     val scala.util.Failure(fh) = Try(Await.result(Future.traverse(Seq(0, 1))(Seq(f1, f2)(_)), Inf))

-    val end = System.nanoTime()
     // Make sure the early failure is always reported, regardless of whether it's on
     // the left or right of the zip/zipWith/sequence/traverse
     assert(fa.getMessage == "Boom Early")
@@ -40,11 +43,6 @@ class FutureTest {
     assert(ff.getMessage == "Boom Early")
     assert(fg.getMessage == "Boom Early")
     assert(fh.getMessage == "Boom Early")
-
-    // Make sure that it fails early, after the earlier failure occurs, and does not wait
-    // for the later failure regardless of which one is on the left and which is on the right
-    assert(end - start > 150 * 1000 * 1000, (end, start))
-    assert(end - start < 950 * 1000 * 1000, (end, start))
   }

   @Test
```

@viktorklang what do you think of this modified test case?

lrytz commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -413,8 +413,14 @@ trait Future[+T] extends Awaitable[T] {
    * @group Transformations
    */
-  def zipWith[U, R](that: Future[U])(f: (T, U) => R)(implicit executor: ExecutionContext): Future[R] =
+  def zipWith[U, R](that: Future[U])(f: (T, U) => R)(implicit executor: ExecutionContext): Future[R] = {
```
The Scaladocs should be updated to reflect the relaxed specification. Also on `Future.zip`.

viktorklang commented:
@lihaoyi I think it's great to avoid the timing component and focus on the order-component as it is going to make for a much more robust test case. Great work!

lihaoyi commented:
Ok, I swapped in the promise-based test case, squashed everything, and copied the PR description into the commit

lrytz reviewed (commented):
LGTM otherwise

lihaoyi commented:
@lrytz I have amended the two scaladocs with the following text:

```
   *  If either input future fails, the resulting future is failed with the same
   *  throwable, without waiting for the other input future to complete.
   *
   *  If the application of `f` throws a throwable, the resulting future
   *  is failed with that throwable if it is non-fatal.
```

viktorklang commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -383,10 +383,11 @@ trait Future[+T] extends Awaitable[T] {
+   *  If either input future fails, the resulting future is failed with the same
+   *  throwable, without waiting for the other input future to complete.
+   *
+   *  If the application of `f` throws a throwable, the resulting future
```
```suggestion
   *  If the application of `f` throws a non-fatal throwable, the resulting future
```

viktorklang commented on test/files/jvm/future-spec/FutureTests.scala:
```diff
@@ -146,7 +146,6 @@ class FutureTests extends MinimalScalaTest {
       assert( ECNotUsed(ec => f.flatMap(_ => fail("flatMap should not have been called"))(ec)) eq f)
       assert( ECNotUsed(ec => f.filter(_ => fail("filter should not have been called"))(ec)) eq f)
       assert( ECNotUsed(ec => f.collect({ case _ => fail("collect should not have been called")})(ec)) eq f)
-      assert( ECNotUsed(ec => f.zipWith(f)({ (_,_) => fail("zipWith should not have been called")})(ec)) eq f)
```
Does this need to be removed now?

lihaoyi commented:
@viktorklang seems not, I put it back in and things pass, though I cant say I'm that familiar with what those tests are asserting

viktorklang commented:
@lihaoyi That test verifies that it does not try to schedule anything on the EC if the originating Future is already failed.

lihaoyi commented:
I think this should be ready for merge unless anyone has any other comments

viktorklang commented on src/library/scala/concurrent/Future.scala:
```diff
@@ -399,11 +400,10 @@ trait Future[+T] extends Awaitable[T] {
+   *  throwable, without waiting for the other input future to complete.
+   *
+   *  If the application of `f` throws a non-fatal throwable, the resulting future
    *  is failed with that throwable if it is non-fatal.
```
```suggestion
   *  is failed with that throwable.
```

lihaoyi replied:
Fixed

viktorklang reviewed (approved):
Great idea, and great work, Li!
